File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
using namespace std;
int dist[102][102];
int dx[4] = {1, 0, -1, 0};
int dy[4] = {0, -1, 0, 1};
int solution(vector<vector<int> > maps){
    queue<pair<int, int>> Q;
    Q.push({0, 0});
    for(int i = 0; i < maps.size(); i++){
        for(int j = 0; j < maps[0].size(); j++){
            if(maps[i][j] == 1) dist[i][j] = -1;
        }
    }
    while(!Q.empty()){
        pair<int, int> cur = Q.front();
        Q.pop();
        dist[0][0] = 1;
        for(int i = 0; i < 4; i++){
            int nx = cur.first + dx[i];
            int ny = cur.second + dy[i];
            if(nx < 0 || nx >= maps[0].size() || ny < 0 || ny >= maps.size()) continue;
            if(dist[ny][nx] >= 0) continue;
            dist[ny][nx] = dist[cur.second][cur.first] + 1;
            Q.push({nx, ny});
        }
    }
    if(dist[maps.size()-1][maps[0].size() - 1] == 0) return -1;
    return dist[maps.size()-1][maps[0].size()-1];
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    int[] dx = {1, 0, -1, 0};
    int[] dy = {0, 1, 0, -1};
    int[][] dist = new int[105][105];
    int[][] board = new int[105][105];
    class Pair {
        int x, y, dist;
        Pair(int x, int y, int dist) {
            this.x = x;
            this.y = y;
            this.dist = dist;
        }
    }
    public int solution(int[][] maps) {
        int height = maps.length;
        int width = maps[0].length;
        for(int i = 0; i < height; i++) {
            for(int j = 0; j < width; j++) {
                board[i + 1][j + 1] = maps[i][j];
            }
        }
        Queue<Pair> queue = new LinkedList<>();
        queue.add(new Pair(1, 1, 1));
        dist[1][1] = 1;
        while(!queue.isEmpty()) {
            Pair cur = queue.poll();
            for(int i = 0; i < 4; i++) {
                int nx = cur.x + dx[i];
                int ny = cur.y + dy[i];
                if(nx < 1 || nx > width || ny < 1 || ny > height) continue;
                if(board[ny][nx] == 0) continue;
                if(dist[ny][nx] != 0 && cur.dist + 1 >= dist[ny][nx]) continue;
                queue.add(new Pair(nx, ny, cur.dist + 1));
                dist[ny][nx] = cur.dist + 1;
            }
        }
        if(dist[height][width] == 0) {
            return -1;
        } else {
            return dist[height][width];
        }
    }
}